backend/src/Tabulators/IRV.ts

```diff
@@ -0,0 +1,241 @@
+import { ballot, candidate, irvResults, irvSummaryData, totalScore } from "./ITabulators";
+
+import { IparsedData } from './ParseData'
+const ParseData = require("./ParseData");
+
+export function IRV(candidates: string[], votes: ballot[], nWinners = 1, breakTiesRandomly = true) {
+    // Determines Instant Runoff winners for given election
+    // Parameters: 
+    // candidates: Array of candidate names
+    // votes: Array of votes, size nVoters x Candidates, zeros indicate no rank
+    // nWiners: Number of winners in election, defaulted to 1 (only supports 1 at the moment)
+    // breakTiesRandomly: In the event of a true tie, should a winner be selected at random, defaulted to true
+
+    // Parse the votes for valid, invalid, and undervotes, and identifies bullet votes
+    const parsedData = ParseData(votes, getIRVBallotValidity)
+
+    const summaryData = getSummaryData(candidates, parsedData)
+
+    // Initialize output data structure
+    const results: irvResults = {
+        elected: [],
+        tied: [],
+        other: [],
+        summaryData: summaryData,
+        logs: [],
+        voteCounts: [],
+        exhaustedVoteCounts: [],
+        overVoteCounts: []
+    }
+
+    let remainingCandidates = [...summaryData.candidates]
+    let activeVotes = parsedData.scores
+    let exhaustedVotes: ballot[] = []
+    let exhaustedVoteCount = 0
+    let overVoteCount = 0
+    while (results.elected.length < 1) { //TODO: multiwinner STV
+        let roundVoteCounts = Array(summaryData.candidates.length).fill(0)
+        for (let i = activeVotes.length - 1; i >= 0; i--) {
+            // loop backwards over the ballots so exhausted votes can be removed
+            let ballot = activeVotes[i]
+            let topRemainingRank: number = 0
+            let topRemainingRankIndex: number = 0
+            let isOverVote = false
+            remainingCandidates.forEach(candidate => {
+                // loop over remaining candidates
+                // find top (non-zero) rank, if duplicate top ranks mark as overvote
+                if (ballot[candidate.index] !== 0) {
+                    // candidate has rank
+
+                    if (topRemainingRank == 0) {
+                        // set initial top rank
+                        topRemainingRankIndex = candidate.index
+                        topRemainingRank = ballot[candidate.index]
+                        isOverVote = false
+                    }
+                    else if (ballot[candidate.index] < ballot[topRemainingRankIndex]) {
+                        // set new top rank
+                        topRemainingRankIndex = candidate.index
+                        topRemainingRank = ballot[candidate.index]
+                        isOverVote = false
+                    }
+                    else if (ballot[candidate.index] === ballot[topRemainingRankIndex]) {
```

Comment:
I think this logic can skip some overvotes. What if ballot[topRemainingRankIndex] is set to 1 on the first iteration and then all remaining ranks are set to 2? This should be an overvote (I think?) but this comparison will evaluate to 1 === 2 for each iteration

Reply:
As said in other comment, overvote is only when its between your top remaining candidates. So if you rank candidate A 1, and rank both candidates B and C 2, it is not an overvote unless A is eliminated. Maybe I'm wrong though.